syncer/syncer.go

```diff
@@ -297,49 +300,78 @@ func (s *Syncer) synchronize(ctx context.Context) bool {
 		log.FieldNamed("processed", s.mesh.ProcessedLayer()))
 
 	s.setStateBeforeSync(ctx)
-	// start a dedicated process for validation.
-	// do not use an unbuffered channel for vQueue. we don't want it to block if the receiver isn't ready. i.e.
-	// if validation for the last layer is still running
-	vQueue := make(chan *types.Layer, s.ticker.GetCurrentLayer().Uint32())
-	vDone := make(chan struct{})
-	s.eg.Go(func() error {
-		s.startValidating(ctx, s.run, vQueue, vDone)
-		return nil
-	})
-	success := false
-	defer func() {
-		close(vQueue)
-		<-vDone
-		s.setStateAfterSync(ctx, success)
-		logger.With().Info(fmt.Sprintf("finished sync run #%v", s.run),
-			log.Bool("success", success),
-			log.String("sync_state", s.getSyncState().String()),
-			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
-			log.FieldNamed("latest", s.mesh.LatestLayer()),
-			log.FieldNamed("processed", s.mesh.ProcessedLayer()))
-		s.setSyncerIdle()
-	}()
 
-	// using ProcessedLayer() instead of LatestLayer() so we can validate layers on a best-efforts basis.
+	var (
+		vQueue chan *types.Layer
+		vDone  chan struct{}
+		// number of attempts to start a validation goroutine
+		attempt = uint64(0)
+		// whether the data sync succeed. validation failure is not checked by design.
+		success = true
+	)
+
+	// using ProcessedLayer() instead of LatestLayer() so we can validate layers on a best-efforts basis
+	// and retry in the next sync run if validation fails.
 	// our clock starts ticking from 1 so it is safe to skip layer 0
 	// always sync to currentLayer-1 to reduce race with gossip and hare/tortoise
+syncLoop:
```

Comment:
i would rearrange this a bit to decouple attempts and queue control from the validation loop, and also remove some duplication

```go
attemptFunc := func() (bool, types.LayerID) {
		var (
			layerID types.LayerID
			layer   *types.Layer
			err     error
		)
		for layerID = s.mesh.ProcessedLayer().Add(1); layerID.Before(s.ticker.GetCurrentLayer()); layerID = layerID.Add(1) {
			if layer, err = s.syncLayer(ctx, layerID); err != nil {
				logger.With().Error("failed to sync to layer", layerID, log.Err(err))
				return false, layerID
			}
			vQueue <- layer
			logger.With().Debug("finished data sync", layerID)
		}
		logger.With().Debug("data is synced, waiting for validation",
			log.Uint64("attempt", attempt),
			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
			log.FieldNamed("latest", s.mesh.LatestLayer()),
			log.FieldNamed("processed", s.mesh.ProcessedLayer()))
		return true, layerID
	}
	// check if we are still on target. if not, go back to the sync loop
	for ; success && target != s.ticker.GetCurrentLayer().Sub(1); attempt++ {
		vQueue, vDone = s.startValidating(ctx, s.run, attempt)
		success, target = attemptFunc()
		close(vQueue)
		select {
		case <-vDone:
		case <-s.shutdownCtx.Done():
			return false
		}
	}
```

Reply:
done

Reply:
thanks for the suggestion. much better now